File: light-control/algorithms/march/march.py
import logging
import math

from .config import config
from .._meta import BaseAlgorithm

class Algorithm(BaseAlgorithm):
    def __init__(self, num_pixels, settings) -> None:
        super().__init__(num_pixels, config, settings)
        self.colors = settings['colors']
        self.cycle = 0
        self.since_change = 0
        self.cycle_time = settings['freq']

        self.set_colors()
    
    def set_black(self, pixel):
        pixel.hue = 0
        pixel.sat = 0
        pixel.val = 0
    
    def set_color(self, pixel, index):
        color = self.colors[index]
        pixel.hue = color[0]
        pixel.sat = color[1]
        pixel.val = color[2]

    def set_colors(self):
        num_colors = len(self.colors)
        count = math.floor(self.cycle / 2)
        is_even_cycle = self.cycle % 2 == 0

        for n in range(len(self.pixels)):
            pixel = self.pixels[n]
            is_even = n % 2 == 0

            if is_even_cycle == is_even:
                self.set_black(pixel)
            else:
                self.set_color(pixel, count)
                count += 1
                if count >= num_colors:
                    count = 0

        self.cycle += 1
        if self.cycle >= num_colors * 2:
            self.cycle = 0
    
    def run_cycle(self, _, elapsed_seconds):
        self.since_change += elapsed_seconds
        if self.since_change > self.cycle_time:
            self.since_change -= self.cycle_time
            self.set_colors()
        
        return super().run_cycle(_, elapsed_seconds)

File: light-control-ts/src/algorithms/march/march.ts
import { BaseAlgorithm, Pixel } from '../_meta/index.js'
import { config } from './config.js'

export class Algorithm extends BaseAlgorithm {
  private readonly colors: number[][]
  private cycle = 0
  private sinceChange = 0
  private readonly cycleTime: number

  constructor(numPixels: number, settings: Record<string, unknown>) {
    super(numPixels, config, settings)
    this.colors = settings['colors'] as number[][]
    this.cycleTime = settings['freq'] as number
    this.setColors()
  }

  private setBlack(pixel: Pixel): void {
    pixel.hue = 0
    pixel.sat = 0
    pixel.val = 0
  }

  private setColor(pixel: Pixel, index: number): void {
    const color = this.colors[index]
    pixel.hue = color[0]
    pixel.sat = color[1]
    pixel.val = color[2]
  }

  private setColors(): void {
    const numColors = this.colors.length
    let count = Math.floor(this.cycle / 2)
    const isEvenCycle = this.cycle % 2 === 0

    for (let n = 0; n < this.pixels.length; n++) {
      const pixel = this.pixels[n]
      const isEven = n % 2 === 0

      if (isEvenCycle === isEven) {
        this.setBlack(pixel)
      } else {
        this.setColor(pixel, count)
        count++
        if (count >= numColors) count = 0
      }
    }

    this.cycle++
    if (this.cycle >= numColors * 2) this.cycle = 0
  }

  runCycle(_elapsedMillis: number, elapsedSeconds: number): boolean {
    this.sinceChange += elapsedSeconds
    if (this.sinceChange > this.cycleTime) {
      this.sinceChange -= this.cycleTime
      this.setColors()
    }
    return super.runCycle(_elapsedMillis, elapsedSeconds)
  }
}
